In [ ]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt



In [ ]:
df = pd.read_csv('laptop.csv')
df

Unnamed: 0.1  Unnamed: 0 Company            TypeName Inches  \
0                0         0.0   Apple           Ultrabook   13.3   
1                1         1.0   Apple           Ultrabook   13.3   
2                2         2.0      HP            Notebook   15.6   
3                3         3.0   Apple           Ultrabook   15.4   
4                4         4.0   Apple           Ultrabook   13.3   
...            ...         ...     ...                 ...    ...   
1298          1298      1298.0  Lenovo  2 in 1 Convertible     14   
1299          1299      1299.0  Lenovo  2 in 1 Convertible   13.3   
1300          1300      1300.0  Lenovo            Notebook     14   
1301          1301      1301.0      HP            Notebook   15.6   
1302          1302      1302.0    Asus            Notebook   15.6   

                                ScreenResolution  \
0             IPS Panel Retina Display 2560x1600   
1                                       1440x900   
2                              Full HD 1920x1080   
3             IPS Panel Retina Display 2880x1800   
4             IPS Panel Retina Display 2560x1600   
...                                          ...   
1298   IPS Panel Full HD / Touchscreen 1920x1080   
1299  IPS Panel Quad HD+ / Touchscreen 3200x1800   
1300                                    1366x768   
1301                                    1366x768   
1302                                    1366x768   

                                       Cpu   Ram               Memory  \
0                     Intel Core i5 2.3GHz   8GB            128GB SSD   
1                     Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
3                     Intel Core i7 2.7GHz  16GB            512GB SSD   
4                     Intel Core i5 3.1GHz   8GB            256GB SSD   
...                                    ...   ...                  ...   
1298            Intel Core i7 6500U 2.5GHz   4GB            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz  16GB            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz   2GB   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz   6GB              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz   4GB            500GB HDD   

                               Gpu       OpSys  Weight        Price  
0     Intel Iris Plus Graphics 640       macOS  1.37kg   71378.6832  
1           Intel HD Graphics 6000       macOS  1.34kg   47895.5232  
2            Intel HD Graphics 620       No OS  1.86kg   30636.0000  
3               AMD Radeon Pro 455       macOS  1.83kg  135195.3360  
4     Intel Iris Plus Graphics 650       macOS  1.37kg   96095.8080  
...                            ...         ...     ...          ...  
1298         Intel HD Graphics 520  Windows 10   1.8kg   33992.6400  
1299         Intel HD Graphics 520  Windows 10   1.3kg   79866.7200  
1300             Intel HD Graphics  Windows 10   1.5kg   12201.1200  
1301            AMD Radeon R5 M330  Windows 10  2.19kg   40705.9200  
1302             Intel HD Graphics  Windows 10   2.2kg   19660.3200  

[1303 rows x 13 columns]

*In this dataset , there are 1303 rows and 13 columns*

In [ ]:
df.columns

Index(['Unnamed: 0.1', 'Unnamed: 0', 'Company', 'TypeName', 'Inches',
       'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight',
       'Price'],
      dtype='object')

In [ ]:
#dropping the irrelevant columns
df = df.drop(['Unnamed: 0.1','Unnamed: 0'],axis=1)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1273 non-null   object 
 1   TypeName          1273 non-null   object 
 2   Inches            1273 non-null   object 
 3   ScreenResolution  1273 non-null   object 
 4   Cpu               1273 non-null   object 
 5   Ram               1273 non-null   object 
 6   Memory            1273 non-null   object 
 7   Gpu               1273 non-null   object 
 8   OpSys             1273 non-null   object 
 9   Weight            1273 non-null   object 
 10  Price             1273 non-null   float64
dtypes: float64(1), object(10)
memory usage: 112.1+ KB


In [ ]:
df.isnull().sum()

Company             30
TypeName            30
Inches              30
ScreenResolution    30
Cpu                 30
Ram                 30
Memory              30
Gpu                 30
OpSys               30
Weight              30
Price               30
dtype: int64

*There are 30 missing values in all columns*

In [ ]:
print(df.duplicated().sum())
print(df.shape)

58
(1303, 11)


*There are 58 duplicate rows in the dataset.*
*There are 1303 rows and 11 columns*

In [ ]:
#drop the duplicate rows
df = df.drop_duplicates()
df.shape

(1245, 11)

In [ ]:
#Checking if the dataset has any null values in it after dropping duplicates
df.isnull().sum()

Company             1
TypeName            1
Inches              1
ScreenResolution    1
Cpu                 1
Ram                 1
Memory              1
Gpu                 1
OpSys               1
Weight              1
Price               1
dtype: int64

In [ ]:
#Since only 1 missing value is present in each column,dropped 1 row
df = df.dropna()
df.isnull().sum()

Company             0
TypeName            0
Inches              0
ScreenResolution    0
Cpu                 0
Ram                 0
Memory              0
Gpu                 0
OpSys               0
Weight              0
Price               0
dtype: int64

In [ ]:
#to check which columns have junk values
junk_values=['?',float('nan'),'']
col = []
for column in df:
  if df[column].isin(junk_values).any():
    col.append(column)

print(col)

['Inches', 'Memory', 'Weight']


*Inches,weight and Memory have junk values*

In [ ]:
df['Memory'].unique()

array(['128GB SSD', '128GB Flash Storage', '256GB SSD', '512GB SSD',
       '500GB HDD', '256GB Flash Storage', '1TB HDD',
       '128GB SSD +  1TB HDD', '256GB SSD +  256GB SSD',
       '64GB Flash Storage', '32GB Flash Storage', '256GB SSD +  1TB HDD',
       '256GB SSD +  2TB HDD', '32GB SSD', '2TB HDD', '64GB SSD',
       '1.0TB Hybrid', '512GB SSD +  1TB HDD', '1TB SSD',
       '256GB SSD +  500GB HDD', '128GB SSD +  2TB HDD',
       '512GB SSD +  512GB SSD', '16GB SSD', '16GB Flash Storage',
       '512GB SSD +  256GB SSD', '512GB SSD +  2TB HDD',
       '64GB Flash Storage +  1TB HDD', '180GB SSD', '1TB HDD +  1TB HDD',
       '32GB HDD', '1TB SSD +  1TB HDD', '?', '512GB Flash Storage',
       '128GB HDD', '240GB SSD', '8GB SSD', '508GB Hybrid', '1.0TB HDD',
       '512GB SSD +  1.0TB Hybrid', '256GB SSD +  1.0TB Hybrid'],
      dtype=object)

In [ ]:
print(df['Inches'].unique())

['13.3' '15.6' '15.4' '14' '12' '17.3' '13.5' '12.5' '13' '18.4' '13.9'
 '11.6' '25.6' '35.6' '12.3' '27.3' '24' '33.5' '?' '31.6' '17' '15'
 '14.1' '11.3' '10.1']


In [ ]:
print(df['Weight'].unique())

['1.37kg' '1.34kg' '1.86kg' '1.83kg' '2.1kg' '2.04kg' '1.3kg' '1.6kg'
 '2.2kg' '0.92kg' '1.22kg' '2.5kg' '1.62kg' '1.91kg' '2.3kg' '1.35kg'
 '1.88kg' '1.89kg' '1.65kg' '2.71kg' '1.2kg' '1.44kg' '2.8kg' '2kg'
 '2.65kg' '2.77kg' '3.2kg' '1.49kg' '2.4kg' '2.13kg' '2.43kg' '1.7kg'
 '1.4kg' '1.8kg' '1.9kg' '3kg' '1.252kg' '2.7kg' '2.02kg' '1.63kg'
 '1.96kg' '1.21kg' '2.45kg' '1.25kg' '1.5kg' '2.62kg' '1.38kg' '1.58kg'
 '1.85kg' '1.23kg' '2.16kg' '2.36kg' '7.2kg' '2.05kg' '1.32kg' '1.75kg'
 '0.97kg' '2.56kg' '1.48kg' '1.74kg' '1.1kg' '1.56kg' '2.03kg' '1.05kg'
 '5.4kg' '4.4kg' '1.90kg' '1.29kg' '2.0kg' '1.95kg' '2.06kg' '1.12kg'
 '3.49kg' '3.35kg' '2.23kg' '?' '2.9kg' '4.42kg' '2.69kg' '2.37kg' '4.7kg'
 '3.6kg' '2.08kg' '4.3kg' '1.68kg' '1.41kg' '4.14kg' '2.18kg' '2.24kg'
 '2.67kg' '4.1kg' '2.14kg' '1.36kg' '2.25kg' '2.15kg' '2.19kg' '2.54kg'
 '3.42kg' '5.8kg' '1.28kg' '2.33kg' '1.45kg' '2.79kg' '8.23kg' '1.26kg'
 '1.84kg' '0.0002kg' '2.6kg' '2.26kg' '3.25kg' '1.59kg' '1.13kg' '1.42kg'
 '1.7

*'?' is present in all the three columns 'Inches','Weight','Memory'*

In [ ]:
#dropping the rows which has '?'
rows_to_drop = df.apply(lambda x:'?' in x.values,axis=1) # Identify rows that have '?' in any column
df = df[~rows_to_drop]# this will remove the '?'

In [ ]:
#Check again if columns have junk values
junk_values=['?',float('nan'),'']
col = []
for column in df:
  if df[column].isin(junk_values).any():
    col.append(column)

print(col)

[]


*None of the columns have junk values. We have removed the rows with the junk values in it.
The dataset is cleaned and have no null values*


In [ ]:
df.dtypes

Company              object
TypeName             object
Inches               object
ScreenResolution     object
Cpu                  object
Ram                  object
Memory               object
Gpu                  object
OpSys                object
Weight               object
Price               float64
dtype: object

#Datatype Transformation

In [ ]:
df['Weight']=df['Weight'].str.replace('kg','').astype(float)
df['Ram'] = df['Ram'].str.replace('GB','').astype(int)
df['Inches'] = df['Inches'].astype(float)


<ipython-input-21-a4a06f210fed>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Weight']=df['Weight'].str.replace('kg','').astype(float)
<ipython-input-21-a4a06f210fed>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Ram'] = df['Ram'].str.replace('GB','').astype(int)
<ipython-input-21-a4a06f210fed>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pand

In [ ]:
df.dtypes

Company              object
TypeName             object
Inches              float64
ScreenResolution     object
Cpu                  object
Ram                   int64
Memory               object
Gpu                  object
OpSys                object
Weight              float64
Price               float64
dtype: object

In [ ]:
df1= df.select_dtypes(include =['object']).columns
df1

Index(['Company', 'TypeName', 'ScreenResolution', 'Cpu', 'Memory', 'Gpu',
       'OpSys'],
      dtype='object')

In [ ]:
df.head(1)

In [ ]:
# plotting countplots for the categorical variables

def drawplot(col):
    plt.figure(figsize=(15,7))
    sns.countplot(x=df[col])
    plt.xticks(rotation='vertical')
    plt.show()

to_view = ['Company', 'TypeName', 'Ram', 'OpSys']
for col in to_view:
    drawplot(col)

Insights:

*   Lenovo,HP, Dell have the highest number of laptops in the market with over 250
*   Type of laptop is 'Notebook' which is maximum
*   Laptop with 8 GB RAM is being sold the most
* Windows 10 has the most market share








In [ ]:
# average price for each of the laptop brands
# this will say us the insight that as per company the price of the laptop vary

plt.figure(figsize=(15,7))
sns.barplot(x = df['Company'],y = df['Price'])
plt.xticks(rotation = 'vertical')
plt.show()

Price of Razer laptop is highest, Gaming laptop price is always on the higher side.

In [ ]:
# various types of laptop

sns.countplot(x=df['TypeName'])
plt.xticks(rotation = 'vertical')
plt.show()

Notebook laptops are most in the market

In [ ]:
# laptop type and variation about the price

sns.barplot(x = df['TypeName'],y = df['Price'])
plt.xticks(rotation = 'vertical')
plt.show()

Workstation has the highest price of around 120000, Gaming laptops ranks 2nd in the price range of 80k-100k

In [ ]:
# variation of inches towards the price

sns.scatterplot(x = df['Inches'],y = df['Price'])

From the above plot, we can say that most of the laptops are within 20inches.
And price of around 17inches have outliers i.e, have highest price

In [ ]:
df['ScreenResolution'].value_counts()

In [ ]:
# creating a new col,Laptop is touchscreen if the value is 1

df['TouchScreen'] = df['ScreenResolution'].apply(lambda x:1
                                                      if 'Touchscreen' in x else 0)

df.head()

In [ ]:
sns.countplot(x= df['TouchScreen'])

In [ ]:
# touch screen on comparison with price of laptop

sns.barplot(x = df['TouchScreen'],y = df['Price'])
plt.xticks(rotation = 'vertical')

In [ ]:
# creating a new col named IPS, does the laptop have IPS facility or not

df['IPS'] = df['ScreenResolution'].apply(
    lambda x:1 if "IPS" in x else 0
)
df.sample(5)

In [ ]:
sns.countplot(x=df['IPS'],palette='plasma')

In [ ]:
# price variation with respect to the IPS col

sns.barplot(x = df['IPS'], y = df['Price'])
plt.xticks(rotation = 'vertical')

# Extracting the X Resolution and the Y Resolution

In [ ]:
'''
We will split the text at the "x" letter and seperate the 2 parts.
From this we can observe that one of the col is Y res, we need to do
some feature engineering on the X res col
'''

df['Screen']=df['ScreenResolution'].str.split().str[-1].str.split('x')
df['x_res'] = df['Screen'].str[0].astype(int)
df['y_res'] = df['Screen'].str[1].astype(int)

In [ ]:
df.head(1)

In [ ]:
df.drop(['Screen'],axis=1,inplace=True)


In [ ]:
df.info()

In [ ]:
#To check the correlation, we need to compare price with only numerical columns,
# so extracting the numerical columns to a dataframe
nv = df[df.select_dtypes(include = ['int32','int64','float32','float64']).columns]
nv

In [ ]:
nv.corr()['Price']

From the correlation plot we observed that as the X_res and Y_res is increasing,the price of the laptop is also increasing,so X_res and Y_res are positively correlated and they are giving much information,so that is the reason why i had splitted Resolution column into X_res and Y_res columns respectively

So to make things good,we can create a new column named PPI{pixels per inch},now as we saw from the correlation plot that the X_res and Y_res are having much collinearity,so why not combine them with Inches which is having less collinearity,so we will combine them as follows,so here is the formula of how to calculate PPI {pixels per inch}



In [ ]:
df['PPI'] = (((df['x_res']**2+df['y_res']**2))**0.5/df['Inches']).astype('float')
df.head()

In [ ]:
nv = df[df.select_dtypes(include = ['int32','int64','float32','float64']).columns]
nv.corr()["Price"]

So we observe from the correlation data that the PPI is having good correlation, so we will be using that, as that is a combination of 3 features and that gives collective results of 3 columns, so we will drop Inches, X_res, Y_res as well

In [ ]:
df.drop(columns=['ScreenResolution', 'Inches','x_res','y_res'],inplace=True)
df.head(1)

# CPU columns extraction

In [ ]:
df['Cpu'].value_counts()

In [ ]:
#So if we observe we need to extract the first 3 words of the CPU column,
#as the first 3 words of every row under the CPU col is the type of the CPU, so we will be using them.

df['Cpu_name']=df['Cpu'].apply(lambda x:" ".join(x.split()[:3]))
df['Cpu_name'].unique()

In [ ]:
'''
As mentioned earlier, if we get any of the Intel versions we will return
them as it is, but if we get any other processer we will first check
whether is that a variant of the intel? or not if yes, then we will tag
it as "Other Intel Processer" else we will say it as "AMD Processer".

'''

def processortype(text):

    if text == 'Intel Core i7' or text == 'Intel Core i5' or text == 'Intel Core i3':
        return text

    else:
        if text.split()[0] == 'Intel':
            return 'Other Intel Processor'

        else:
            return 'AMD Processor'

df['Cpu_name'] = df['Cpu_name'].apply(lambda text:processortype(text))
df.head()

In [ ]:
df.drop(['Cpu'],axis=1,inplace=True)

In [ ]:
sns.countplot(x=df['Cpu_name'],palette='viridis')
plt.xticks(rotation = 'vertical')
plt.show()

Laptop with Intel Core i7 processor has the highest count in the market

In [ ]:
# price vs processer variation

sns.barplot(x=df['Cpu_name'],y=df['Price'],palette = 'viridis')
plt.xticks(rotation = 'vertical')
plt.show()

Intel Core i7 also costs more than the rest of the processor

In [ ]:
df.head(1)

# RAM Analysis

In [ ]:
sns.countplot(x=df['Ram'])

In [ ]:
sns.barplot(x=df['Ram'],y=df['Price'])

Price increases as the RAM increase. So, RAM has a +ve coorelation with price


# Extract Memory type from Memory Column

In [ ]:
df['Memory'].value_counts()

In [ ]:
#We will seperate the Type of memory and the value of it.There are different dimensions in the data,128GB SSD + 1TB HDD,
#so inorder to ofr it come in a same dimension we need to do some modifications which are done below as shown

# 4 most common variants observed : HDD,SDD,Flash, Hybrid
# this expression will remove the decimal space for example 1.0 TB will be 1TB
df['Memory'] = df['Memory'].astype(str).replace('\.0','',regex = True)

# replace the GB word with ""

df['Memory'] = df['Memory'].str.replace('GB','')

# replace the TB word with "000"

df['Memory'] = df['Memory'].str.replace('TB','000')

# split the word across the "+" character

new_df = df['Memory'].str.split("+",n=1,expand = True)
new_df


In [ ]:
# we will strip up all the white spaces, basically eliminating white space

df['first'] = new_df[0]
df['first'] = df['first'].str.strip()
df.head()

# HDD

In [ ]:
def applychanges(value):
    df['Layer1'+value] = df['first'].apply(lambda x:1 if value in x else 0)

list_to_apply = ['HDD', 'SSD', 'Hybrid', 'FlashStorage']
for value in list_to_apply:
    applychanges(value)

df.head()

In [ ]:
# remove all the characters just keep the numbers
df['first'] = df['first'].str.replace(r'\D', '', regex=True)
df['first'].value_counts()

In [ ]:
df['Second'] = new_df[1]
df.head()
df['Second'].unique()

In [ ]:
def applychanges1(value):
    df['Layer2'+value] = df['Second'].apply(lambda x:1 if value in x else 0)

list_to_apply1 = ['HDD', 'SSD', 'Hybrid', 'FlashStorage']
df['Second'] = df['Second'].fillna("0")
for value in list_to_apply1:
    applychanges1(value)

# remove all the characters just keep the numbers

df['Second'] = df['Second'].str.replace(r'\D','',regex=True)
df['Second'].value_counts()

In [ ]:
df['first'] = df['first'].astype('int')
df['Second'] = df['Second'].astype('int')
df.head()

In [ ]:
df["HDD"]=(df["first"]*df["Layer1HDD"]+df["Second"]*df["Layer2HDD"])
df["SSD"]=(df["first"]*df["Layer1SSD"]+df["Second"]*df["Layer2SSD"])
df["Hybrid"]=(df["first"]*df["Layer1Hybrid"]+df["Second"]*df["Layer2Hybrid"])
df["Flash_Storage"]=(df["first"]*df["Layer1FlashStorage"]+df["Second"]*df["Layer2FlashStorage"])


In [ ]:
# droping of unnecessary columns

df.drop(columns=['first', 'Second', 'Layer1HDD', 'Layer1SSD', 'Layer1Hybrid',
                'Layer1FlashStorage', 'Layer2HDD', 'Layer2SSD', 'Layer2Hybrid',
                'Layer2FlashStorage'],inplace=True)

In [ ]:
df.drop(columns=['Memory'],inplace=True)
df.sample(5)

In [ ]:
df.head(1)

# GPU Analysis

In [ ]:
df['Gpu'].value_counts()

In [ ]:
df.loc[1]['Gpu'].split()[0]

In [ ]:
df['Gpu_brand'] = df['Gpu'].apply(lambda x: x.split()[0])
df['Gpu_brand'].unique()

In [ ]:
sns.countplot(x=df['Gpu_brand'])

In [ ]:
df = df[df['Gpu_brand']!='ARM']
sns.countplot(x=df['Gpu_brand'])

In [ ]:
sns.barplot(x=df['Gpu_brand'],y=df['Price'])

Nvidia Gpu brand has most price compared to Intel and AMD

In [ ]:
df = df.drop(columns=['Gpu'])
df.head(1)

In [ ]:
DF_copy = df

# Operating System

In [ ]:
df['OpSys'].value_counts()

In [ ]:
def setcategory(text):

    if text == 'Windows 10' or text == 'Windows 7' or text == 'Windows 10 S':
        return 'Windows'
    elif text == 'Mac OS X' or text == 'macOS':
        return 'Mac'
    else:
        return 'Other'

df['OpSys'] = df['OpSys'].apply(lambda text:setcategory(text))



In [ ]:
sns.countplot(x=df['OpSys'])

In [ ]:
sns.barplot(x=df['OpSys'],y=df['Price'])

Price of Mac OS is highest

# Weight Analysis

In [ ]:
sns.histplot(df['Weight'])

In [ ]:
sns.scatterplot(x=df['Weight'],y=df['Price'])

Weight increases, the price also tends to increase, at least up to a certain point.
There does not seem to be a clear linear relationship as the points do not form a distinct line.

# Price Analysis

In [ ]:
sns.histplot(df['Price'])

In [ ]:
df['Price']=np.log(df['Price'])


In [ ]:
sns.boxplot(df,y='Price')

In [ ]:
df.dtypes

In [ ]:
nv = df[df.select_dtypes(include = ['int32','int64','float32','float64']).columns]
nv

In [ ]:
nv.corr()['Price']

In [ ]:
plt.figure(figsize=(10,5))
sns.heatmap(nv.corr(), annot=True,cmap='plasma')



*  Strong positive correlation with Price (0.63): Higher RAM likely leads to higher prices.
*  Strong positive correlation with SSD (0.65): Higher-priced devices often come with SSDs.
*  Moderate positive correlations with PPI (0.47) and IPS (0.28): Higher-priced devices may have better screens.

In [ ]:
df.head()

In [ ]:
df.drop(columns = ['Hybrid','Flash_Storage'],axis=1,inplace=True)

In [ ]:
df.head(1)

# Now we divide the train and test data

In [ ]:
X = df.drop(['Price'],axis=1)
y= df['Price']


In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder

from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_absolute_error,r2_score
from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn import tree
import warnings
warnings.filterwarnings('ignore')


In [ ]:
X.head()

In [ ]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
X_train.shape,X_test.shape

In [ ]:
X_test

In [ ]:
mapper = {i:value for i, value in enumerate(X_train.columns)}
mapper

# Linear Regression

In [ ]:
step1 = ColumnTransformer(transformers=[
    ('col_tnf', OneHotEncoder(sparse=False, drop=None,handle_unknown = 'ignore'),[0,1,3,8,11])
],remainder='passthrough')

step2 = LinearRegression()

pipe = Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)

y_pred = pipe.predict(X_test)

print('R2 score',r2_score(y_test,y_pred))
print('MAE',mean_absolute_error(y_test,y_pred))

# Decision Tree

In [ ]:
step1 = ColumnTransformer(transformers=[
    ('col_tnf', OneHotEncoder(sparse=False, drop=None,handle_unknown = 'ignore'),[0,1,3,8,11])
],remainder='passthrough')

step2 = DecisionTreeRegressor(max_depth=8)

pipe = Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)

y_pred = pipe.predict(X_test)

print('R2 score',r2_score(y_test,y_pred))
print('MAE',mean_absolute_error(y_test,y_pred))

# Random Forest

In [ ]:
step1 = ColumnTransformer(transformers=[
    ('col_tnf', OneHotEncoder(sparse=False, drop=None,handle_unknown = 'ignore'),[0,1,3,8,11])
],remainder='passthrough')

step2 = RandomForestRegressor(n_estimators=100,
                              random_state=3,
                              max_samples=0.5,
                              max_features=0.75,
                              max_depth=15)

pipe = Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)

y_pred = pipe.predict(X_test)

print('R2 score',r2_score(y_test,y_pred))
print('MAE',mean_absolute_error(y_test,y_pred))



# Gradient Boosting

In [ ]:
step1 = ColumnTransformer(transformers=[
    ('col_tnf', OneHotEncoder(sparse=False, drop=None,handle_unknown = 'ignore'),[0,1,3,8,11])
],remainder='passthrough')

step2 = GradientBoostingRegressor(n_estimators=500)

pipe = Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)

y_pred = pipe.predict(X_test)

print('R2 score',r2_score(y_test,y_pred))
print('MAE',mean_absolute_error(y_test,y_pred))

In [ ]:
X_train.values

In [ ]:
import pickle

pickle.dump(df,open('df_boost.pkl','wb'))
pickle.dump(pipe,open('pipe_boost.pkl','wb'))


# XG Boost

In [ ]:
step1 = ColumnTransformer(transformers=[
    ('col_tnf', OneHotEncoder(sparse=False, drop=None,handle_unknown = 'ignore'),[0,1,3,8,11])
],remainder='passthrough')

step2 = XGBRegressor(n_estimators=45,max_depth=5,learning_rate=0.5)
pipe = Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)

y_pred = pipe.predict(X_test)

print('R2 score',r2_score(y_test,y_pred))
print('MAE',mean_absolute_error(y_test,y_pred))

Questions:

Which features have the most significant impact on laptop prices?

-- Based on the coorelation matrix that we have, we see that Higher RAM likely leads to higher prices.
SSD's also have quite a good impact on the prices after RAM
Higher-priced devices often come with SSDs.

Can the model accurately predict the prices of laptops from lesser-known brands?

--Yes.

Does the brand of the laptop significantly influence its price?

--No, brand of the laptop doesnt have much significance on the price


How well does the model perform on laptops with high-end specifications compared to budget laptops?

--Price predicted by the model is 70% with high end specifications


What are the limitations and challenges in predicting laptop prices accurately?

--Prices can become outdated quickly, so integrating real-time data is crucial but challenging.
Market Volatility: Prices change frequently due to demand, supply issues, and new model releases.
Geographical Variations: Prices vary in different regions due to taxes, tariffs, and consumer preferences.

How does the model perform when predicting the prices of newly released laptops not present in the training dataset?

--the model has been trained primarily on data from older laptop models, it wont accurately predict prices for new releases, especially if these new laptops feature technologies or specifications not previously captured in the training data.
